### DISCUSSION TITLE
text extraction for second end-to-end iteration

### DISCUSSION THREAD

**Discussion by oepen** [hplt-project/HPLT-textpipes#1] [Ideas]
There has been extensive discussion between the Prompsit and UiO teams about the technical setup for our second round of data preparation.  We will summarize and collectively discuss here, aiming for decisions by the end of January.

## Functionality Overview: Stage 1 (Currently `warc2text`)

+ WARC parsing
  + metadata extraction, e.g. URL
  + *new*: more metadata, e.g. capture time, content type, byte offset
+ record selection: natural language text (or `robots.txt`)
  + can include e.g. URL-based filtering
  + *new*: collect all `robots.txt` to new archive
+ HTTP and HTML metadata extraction, e.g. language cues
+ encoding detection and normalization
  + everything should be UTF-8 (though may contain invalid code points)
+ (sequential) HMTL reading
  + text extraction (maybe HTML character entity normalization)
  + whitespace normalization
  + segmentation by among others block-level elements
  + *new*: segment-introducing elements as metadata
+ language identification
  + document-level
  + segment-level
+ output generation
  + one document per line
  + "tabular storage" text and each piece of metadata in separate, line-aligned files
    + allows reading (and extending) subsets of data independent of the rest
    + not HPC-friendly: heavy load on filesystem; reduced locality of related data
  + text BASE64-encoded, to escape newlines and protect against invalid encoding
    + [relatively heavy overhead in terms of data volume](https://github.com/bitextor/warc2text/issues/34)
  + alternate output format, writing everything into one JSONlines file

## New Tool: Stage 2

To filter based on `robots.txt`-compliance, we need to insert an additional step: Stage 1 collects all `robots.txt` across WARC files and crawls; to flag (or remove) documents based on this, we need to compare all URLs against this new mega-index.  This applies to mono- and bi-texting alike and may substantially remove the overall volume of data, so should be applied prior to downstream processing.

Unrelated to the above, we are tempted to put more functionality into text extraction, notably boilerplate detection (or removal), e.g. using a tool like [Trafilatura](https://trafilatura.readthedocs.io/en/latest/index.html).  Doing this prior to (document-level) language identification, will likely lead to more consistent language identification.  Therefore, this would also have to go into Stage 2 then.

A major benefit of breaking text extraction into two stages is that Stage 1 can be lean and cheap, so that it can be run directly on the storage nodes.  Record selection from the WARC files (skipping e.g. binary files, images, videos, non-HTML documents) massively reduces the data volume, so that the resulting input to Stage 2 is a lot easier to transfer to a compute cluster.  Running Stage 1 on a real cluster, in turn, means we can be more ambitious in processing, e.g. afford full HTML DOM parsing to support boilerplate detection (and possibly better segmentation).  It also means we can (probably) implement Stage 2 in Python (rather than in C++), where HPLT has a larger developer team available.

For this scenario, Stage 1 (`warc2text`) needs to be refined (into something that could be called `warc2html`) to write out complete HTML documents, while skipping text extraction, segmentation, and language identification (also, it appears that encoding normalization currently is only applied during text extraction, so it would have to be moved slightly earlier inside `warc2text`).  Abstractly, its output should be complete HTML documents and associated metadata.

In this revised scenario, we are free to define Stage 1 output independent of user existing use cases for `warc2text`; for now at least, this mode would only be used as part of the revised HPLT text extraction pipeline. Two options have been discussed:

(1) A conservative choice would be to moderately adjust existing output code in `warc2text` and keep the HTML documents separate from the metadata, i.e. create pairs of line-aligned files: one with BASE64-encoded HTML documents, the other with all metadata in JSONlines.  The rationale for the separation is volume.  HTML documents likely account for the by far biggest part of the data; being able to just process metadata (e.g. for analytics, or `robots.txt`-based filtering) without having to read through all the textual content may be beneficial.  On the other hand, integrating all metadata into a single JSONlines file will make the pipeline more HPC-friendly and practically make it easier to deal with groups of metadata fields.

(2)

**Comment by ZJaume** [comment]
Even if we just write the html, the text extraction in warc2text is required for the [tag filters](https://github.com/hplt-project/warc2text-runner/blob/main/mt-filter-list.annotated) that remove google translated wordpress documents. Are you ok doing that kind of filtering in stage 2? It's just a matter of detecting certain html tags and could be easily done with xpath.
@oepen @nvanva @akutuzov

**Reply by nvanva** [reply]
Looks nice to me. Maybe you could even annotate them instead of filtering to be consistent with our general approach.

**Reply by oepen** [reply]
I think we should follow conceptual design principles and move functionality where it belongs: inspection of the HTML content, including for pattern detection on MT-introduced DOM structure, would seem to belong in Stage 2.  And not running the text extraction part in the revised `warc2html` should help to make that process lean and efficient on the storage nodes.

**Comment by ZJaume** [comment]
Regarding the output format, at Prompsit we are discarding any use of binary indexed formats, since it will complicate things in warc2text to much more development time than we want to expend before starting the extraction. So then, we have two options left: all the {meta}data in jsonl or only metadata in jsol and html in b64. Despite I agree with the fact that is very easy to perform metadata-related queries when it is separated, I think it will complicate things much more for the stage 2 parallelization. If we run, let's say, 256 jobs of warc2text, we will have 256 jsonl file + 256 b64 files. Given that stage 2 will need more parallelization than running 256 python processes, parallelization within each file using GNU parallel would be needed. But I don't think that's an easy solution if each document has to be read from two sources (metadata, html). When for example, with a single jsonl file it would be as easy as:
```
zcat text/wide16/batch1.jsonl.gz | parallel --pipe -j64 -k python stage2-processing.py
```

**Reply by oepen** [reply]
I think whether the data stream is represented in one file or in two line-aligned files should have little to say for parallelization in Stage 2, even though I see your point that just reading from standard input would not be an option.  Instead, the Stage 2 tool could be given two files names as arguments, create two streams and read from them in a line-oriented loop.

Python has "built-in" support for reading a variety of compressed formats, and these may or may not spawn a separate process for the decompression.  I expect avoiding a separate process could be beneficial, as there is overhead in "pumping" all data through a pipe – essentially making a full but unnecessary copy, as the underlying write(2)s and read(2)s move the data from one memory block to another one.  So, it is possible avoiding passing the data via standard input can be beneficial.

I can also imagine that parallelization using (lean) Posix threads may give better efficiency than using full-blown processes, though the Python GIL issue may still present obstacles here.  But either way, I would weigh the conceptual benefit of the hybrid, line-aligned format (BASE64 and JSONlines) stronger than the (undeniable) appeal of designing the Stage 2 tool like a classic Un*x pipeline tool.  I imagine we will soon enough find ourselves doing interesting things with just the metadata.

**Reply by nvanva** [reply]
When running warc2text we have 256 parallel processes, but each gets a batch of 100 WARCs, so in total there are much more tasks and hence output folders (e.g. WIDE17 has ~660 000 WARCs resulting in ~6600 tasks and output folders). After getting rid of language identification in warc2text we will no longer split texts into as many subfolders as there are languages, this will result in 200-300x less files generated and we can reduce batch size and generate even more output folders if needed for further parallelization. So  I am also in doubt if we really want to parallelize across lines in addition to parallelizing across files.

**Reply by nvanva** [reply]
Maybe you already know that, but an interesting thing I've learnt when working with warc2text outputs is that you can simply and very efficiently create a TSV from separate gziped columns:
```paste <(zcat url.gz) <(zcat date.gz) <(zcat text.gz)```
But I'm not sure yet how to do that if you have one of the files in json format with several columns.

**Reply by akutuzov** [reply]
> But I'm not sure yet how to do that if you have one of the files in json format with several columns.

`paste <(zstdcat eo_1.jsonl.zst | jq -r .collection) <(zstdcat eo_1.jsonl.zst | jq -r .document_lang)`
:-)

**Reply by nvanva** [reply]
Looks like a solution :) But probably not the most efficient since we the same file eo_1.jsonl.zst is uncompressed with zstdcat and parsed with jq twice.

**Reply by jelmervdl** [reply]
`zstdcat en_1.jsonl.zst | jq -r "[.collection,.document_lang]|@tsv"`

Not sure if this is the most efficient way, but it does generate tsv.

**Comment by ZJaume** [comment]
All the changes that we've been talking about are made now in [#50](https://github.com/bitextor/warc2text/pull/50). But I made a couple of changes more, in separated PRs because we may not merge them. These changes are:
- https://github.com/bitextor/warc2text/pull/51 to write files in this format instead of GZIP. This could allow us, either to speed up the warc2text execution (and have a little bit more of compression ratio) by just changing to ZSTD or, to have a similar speed than GZIP but to obtain much higher compression ratio using ZSTD and level 9 (instead of the default 3).
- https://github.com/bitextor/warc2text/pull/52. This now requires `--jsonl --stdout` to write all the output format in a JSONL to stdout that Jelmer did. Then, using only `--jsonl` changes the format of the `text.gz` and `html.gz` files (we don't need the `text.gz`, but just for consistency) from base64, one document per line, to JSON, one document per line.

I made these two new changes because we probably may need to cut down the output size significantly. As we talked about in a discussion in Mattermost, we don't have much space on NIRD and the extraction of HTML instead of text decreases much less the size of the WARCS. About 1/3 or 1/2 compared to 10x-30x times as before extracting text. As an example, 50GB of random WARCs yielded 16GB of html.gz and 2.9GB of text.gz. With the changes above mentioned, taking advantage of ZSTD compression and the JSON plain text format we can obtain much higher compression ratios (as we early discussed b64 is compressing less than plain text).

You may notice that the changes are made in form of a combination of pluggable features. This is because I wanted to include and maintain the changes in the main repo and not having them splitted between main and fork. So I tried to make all of them in a way they don't break the internal design and don't change the default behaviour, but let us in HPLT manage the output format more or less as we need.

EDIT: Please if you want to try all the changes, use https://github.com/bitextor/warc2text/pull/52. As the PR are made incrementally and that one includes all the features.

**Reply by ZJaume** [reply]
An example command that we may use for Stage 1, I think it would be:
```bash
warc2text \
  -f html,metadata --jsonl \
  --compress zstd --compress-level 9 \
  --skip-extraction --classifier skip \
  --url-filters url-filter-list.optimised \
  -o html/ hplt/data/**.warc.gz
```

**Reply by ZJaume** [reply]
That would write a single directory `html/` containing `html.zst` and `metadata.jsonl.zst`

**Reply by ZJaume** [reply]
With the example of 50 random WARCs above mentioned that yielded 16GB of `html.gz`, writing the same HTML still encoded in b64 but with zstd -9 leads to a `html.zst` of 11GB. That same compression but using JSONL instead of b64, leads to a 6.2GB file. That is a 1/8 reduction instead 1/3.

**Reply by nvanva** [reply]
This works for me after replacing --skip-extraction with --skip-text-extraction. It took me 35m to process 1% of wide15, which was previously processed in 23m with cld2 and 46m with fastertext. And the size of the outputs is now 557 GB compared to 215 GB before.

So should we probably merge these changes to master and create the new version of warc2text (to install with EasyBuild on all preprocessing nodes), then start processing all the downloaded crawls?
1) What size of a batch (number of WARCs fed to a single warc2text job) do we prefer? For the first iteration bs=100  was used. I remember @ZJaume mentioned he wants smaller batches for better parallelization, it this correct?
2) Should we also add --robotspass flag to save robots.txt for all crawls in this round?

**Reply by ZJaume** [reply]
I suggested smaller batch for parallelization of Stage2, if there wasn't gonna have line-based parallelization within each file. But if it's going to be, I don't think we need to change that.

**Comment by ZJaume** [comment]
A new release has been made with the changes. If there is nothing more needed, I guess processing can start.
https://github.com/bitextor/warc2text/releases/tag/v1.2.0
